### DISCUSSION TITLE
Having an option to do AND operation with ConditionalAttribute

### DISCUSSION THREAD

**Discussion by gustavopsantos** [dotnet/csharplang#6896] [Language Ideas]
Currently, ConditionalAttribute defines the attribute `[System.AttributeUsageAttribute(System.AttributeTargets.Class | System.AttributeTargets.Method, AllowMultiple=true)]` with `AllowMultiple=true` which allows us to do OR operation.
But also it would be helpful to have a way of doing AND operation as well.
Example:
In unity we have a field for setting Scripting Define Symbols, and sadly those symbols are used in both Editor and Build.
And its a bit sad, because in some cases, we develop some heavy debugging code that we want to run only on Unity Editor but also allowing to disable the debugging for perf profilling purposes.
So I already saw myself needing to do the following:

```csharp
[Conditional("UNITY_EDITOR")] and [Conditional("MY_TOOL_DEBUG")]
public void DoSomeReflectionHeavyDebuggingStuff()
{
} 
```

So with this I have the following benefits:
- Be sure that this code never gets in builds
- Im still able to profile the performance of the tool in unity editor, without the having the debug options enable by undefining `MY_TOOL_DEBUG`

**Comment by HaloFour** [comment]
You can accomplish this today by defining additional symbols only if the requisite symbols are defined:

```cs
#if UNITY_EDITOR && MY_TOOL_DEBUG
    #define UNITY_EDITOR_AND_MY_TOOL_DEBUG
#endif
```

then

```cs
[Conditional("UNITY_EDITOR_AND_MY_TOOL_DEBUG")]
public void DoSomeReflectionHeavyDebuggingStuff()
{
} 
```

**Comment by theunrepentantgeek** [comment]
You can effectively do this already, just by definition an extra preprocessing symbol.

``` csharp
// At the top of the file
#if UNITY_EDITOR && MY_TOOL_DEBUG
#define DO_SOME_REFLECTION_HEAVY_DEBUGGING_STUFF
#endif

// later on
[Conditional("DO_SOME_REFLECTION_HEAVY_DEBUGGING_STUFF")]
private static void DoSomeReflectionHeavyDebuggingStuff() 
{
   Console.WriteLine("Heavy!");
}
```

Try out [this SharpLab example](https://sharplab.io/#v2:C4LgTgrgdgPgsAKAMQBMCmAzAllNACAVQDkBJAFQE0B9AUQBFyB5AJUVUx3wFlqzHGAMlTo0AQgQDiiNlgyFSlWgz7M8AMjV4eVPoOFjJbdNlx46jKgGVGXGlWY0AYgJoBhMiUZEqACRoBBADVqEXEJCRIiCSsyAkdHNjQoFFlpBAABACYARkR0gAY8dOyAOjosAEMAcygAewBnYCwAY3qAbjT0gGYizLxXPABvRDxR3r7hhDHpouyATgAKACISYAByerwKs3wBWrBKqCWASg6pmdG6WstagFs0ZkwAGzRmptqoHzQKgDcATzoaAARhAqlUcFVLMAIBgMAtTiMLrNFksJGBvsA8JZmrVgMATmdpgBfRGjNLTADarg+KXeUAqT2W5isNjsDmcbg8Xl8AWC+jCESiMTijhOAF1SXgAA4HH4VYD4YoANiKABYzNc7g9nq86V9fgDgaDwVBIdDYfC8JLJkjiij9f8AIQEyUkhBEoA===)

When you're looking at the output, remember that the compiler omits _calls_, not the method itself. Easiest way to see the difference in my example is you looking at the Run results.

**Comment by gustavopsantos** [comment]
This method works only if the usage of the method defining the ConditionalAttribute is in the same file as the #define.
I wanted to be able to define a conditional method implementation in a abstract class, with the #define in this abstract class. While the inheritors would not have to mess with

```c#
#if UNITY_EDITOR && MY_TOOL_DEBUG
#define DO_SOME_REFLECTION_HEAVY_DEBUGGING_STUFF
#endif
```

So explaining a bit more
I have this Resolver abstract class:
```c#
#if UNITY_EDITOR && MY_TOOL_DEBUG
#define DO_SOME_REFLECTION_HEAVY_DEBUGGING_STUFF
#endif

internal abstract class Resolver
{
    public abstract object Resolve(Container container);

    [Conditional("DO_SOME_REFLECTION_HEAVY_DEBUGGING_STUFF")]
    protected void DoSomeReflectionHeavyDebuggingStuff()
    {
        UnityEngine.Debug.LogError($"Running {GetType().Name}");
    }
}
```

```c#
internal sealed class ValueResolver : Resolver
{
    private readonly object _value;

    public InstanceResolver(object value)
    {
        _value = value;
    }

    public override object Resolve(Container container)
    {
        DoSomeReflectionHeavyDebuggingStuff();
        return _value;
    }
}
```

And I have multiple classes inheriting from Resolver, I do not wanted to #define in every inheritor, it looks counterintuitive and error-prone

**Comment by gustavopsantos** [comment]
So adding a bit more.
When I add symbols to Unitys Scripting Define Synbols, unity simply goes and updates the `<DefineConstants>` xml entry in the .csproj file.
So defining symbols at `Unitys Scripting Define Synbols` defines it globally.
While defining symbols at file, are local.
They are not equivalent.

**Comment by HaloFour** [comment]
If there was a change here I think I would prefer if `ConditionalAttribute` would have a new constructor that could accept `params string[]`.  Then you could apply the attribute as follows:

```cs
[Conditional("FOO", "BAR")]
```

Or maybe a suggestion of a way to declare global preprocessor symbols makes sense.

As for the suggested syntax, I don't think that the `and` keyword makes sense given attributes are a general purpose language feature.

**Reply by gustavopsantos** [reply]
Sure, I agree, I put an `and` there just to make more visible what I wanted, but yeah, I was thinking about:
- Or changing the constructor to accept `params string[]`
- Or ading an overload accepting `string[]`

**Comment by gustavopsantos** [comment]
@HaloFour I saw that you already commited to this repository. How it works? Should I propose this API on a issue? Thanks,

**Comment by gustavopsantos** [comment]
@stephentoub I saw that before proposing it, we should create a issue here, and get feedback first.
What do you think about it? Can I make a proposal about it?

**Reply by stephentoub** [reply]
This isn't first and foremost about an attribute; rather, it's about the language/compiler's ability to do Boolean logic on conditional compilation of method call sites.  It would be driven as a language/compiler feature, and then if agreed upon, we'd expose from the runtime/libraries whatever attribution mechanism was agreed upon to enable the language/compiler.  In other words, any proposal here would be about the language/compiler feature rather than about an attribute, which is just a detail of the larger feature.

@333fred, would this be something that would need to be considered by LDM, or is this "just" a compiler feature request?

**Reply by 333fred** [reply]
It would need to be a language request, as the behavior of `Conditional` is specified in the language: https://github.com/dotnet/csharpstandard/blob/b76afdc244e4b750488da6f35220c11ec5de7186/standard/attributes.md#22532-conditional-methods.

That being said, you could write a simple msbuild property to combine multiple flags and define across the entire compilation, if you don't like the `#if` version that was shared above.
